### PR TITLE
Robot controls for drive team

### LINKED ISSUE #30: Verify usage of getCurrentLiftLevel and getDestinationLiftLevel.

### LINKED ISSUE #32: Add controls that drive team wants
Controls to be implemented are below. They can be checked when implemented.

## Driver
- [x] Right Y axis is forward/backward
- [x] Left X axis is turn left/right
- [x] Right trigger to collect prism (runs intake and grabbers in). Overrides manipulator collection control, but not spin.
## Manipulator
- [x] Left bumper timedSpin, left intake motor pulls more than the right motor
- [x] Right bumper timedSpin, right intake motor pulls more than the left motor
- [x] Left joystick Y axis combined intake and grabber move controls, down for cube in
- [x] Left trigger > 0.5 bulldoze mode, intakes out at full speed, grabber uneffected
- [x] D-pad up max grabber eject
- [x] D-pad left medium grabber eject
- [x] D-pad down minimum grabber eject
- [x] Right joystick Y axis lift control, down for raise lift
- [x] Y button lift level up, `getDestinationLevel().next()`
- [x] A button lift level down, `getDestinationLevel().previous()`
- [x] Right trigger > 0.5 use manual mode to move lift. (Right joystick sets motor power) Upon release the PID controller setpoint is the current lift height.
- [x] X button currently resets the lift encoder. Should be disabled.
### Unused manipulator inputs
- [x] B button
- [x] Left stick click
- [x] Right stick click
- [x] Back button
- [x] Start button

### PR DESCRIPTION
**Pull request by docprofsky**
Closes #32, resolves to resolve #30.
Works towards deborkifying #34.

### CODE REVIEW

File: src/com/teamcautionrobotics/misc2018/FunctionRunnerSendable.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package com.teamcautionrobotics.misc2018;
+
+import java.util.function.BooleanSupplier;
+
+import edu.wpi.first.wpilibj.Sendable;
+import edu.wpi.first.wpilibj.SendableBase;
+import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
+
+public class FunctionRunnerSendable extends SendableBase implements Sendable {
```
Comment (TaeSeanDo) [inline src/com/teamcautionrobotics/misc2018/FunctionRunnerSendable.java]: Should include a javadoc to explain the purpose of this class.

File: src/com/teamcautionrobotics/robot2018/Robot.java [hunk 2]
```diff
@@ -163,43 +179,88 @@ public void teleopPeriodic() {
...
             }
             liftLowerButtonPressed = liftLowerButton;
         }
 
-        if (manipulator.getButton(Button.X)) {
-            lift.resetEncoder();
+        boolean liftPIDManualModeTrigger = manipulator.getAxis(Axis.RIGHT_TRIGGER) > 0.5;
+
+        if (!liftPIDManualModeTrigger && liftPIDManualModeTrigger != liftPIDManualModeEnabled) {
+            lift.setHeight(lift.getCurrentHeight());
```
Comment (TaeSeanDo) [inline src/com/teamcautionrobotics/robot2018/Robot.java]: setHeight should probably be renamed to setHeightPID.

File: src/com/teamcautionrobotics/robot2018/Robot.java [hunk 3]
```diff
@@ -70,6 +75,15 @@ public void robotInit() {
         intake = new Intake(3, 4, 5);
         lift = new Lift(2, 4, 5, 0.8, 0.1, 0.4);
 
+        liftResetSendable = new FunctionRunnerSendable("Reset lift", () -> {
```
Comment (TaeSeanDo) [inline src/com/teamcautionrobotics/robot2018/Robot.java]: This should probably be renamed to liftResetEncoderSendable; put "reset lift encoder" as the string as well.

File: src/com/teamcautionrobotics/robot2018/Robot.java [hunk 4]
```diff
@@ -151,9 +165,11 @@ public void teleopPeriodic() {
...
 
+        boolean driverPrismHarvesting = false;
+        double grabberPower = 0;
+        double intakePower = 0;
         if (lift.getCurrentHeight() < MAX_LIFT_HEIGHT_FOR_INTAKE) {
-            intake.move(manipulator.getAxis(Axis.LEFT_Y));
-
+            // Spin controls only permitted if lift is down
             // Left bumper spins counterclockwise
             if (manipulator.getButton(Button.LEFT_BUMPER)) {
```
Comment (TaeSeanDo) [inline src/com/teamcautionrobotics/robot2018/Robot.java]: Spinning should also be disabled if driverPrismHarvesting.

File: src/com/teamcautionrobotics/robot2018/Robot.java [hunk 5]
```diff
@@ -163,43 +179,88 @@ public void teleopPeriodic() {
...
+
+                    // D-pad is down
+                    case 180:
+                        grabberPower = -0.1;
+                        break;
+                }
+            }
+        }
+
+        intake.move(grabberPower, intakePower);
```
Comment (TaeSeanDo) [inline src/com/teamcautionrobotics/robot2018/Robot.java]: This may conflict with a timed spin or a bulldoze by resetting motor values.

**Review by TaeSeanDo — CHANGES_REQUESTED** [review]

File: src/com/teamcautionrobotics/robot2018/Robot.java (hunk 2 above)
Comment (docprofsky) [reply src/com/teamcautionrobotics/robot2018/Robot.java]: How about `setDestinationHeight`?

File: src/com/teamcautionrobotics/robot2018/Robot.java (hunk 5 above)
Comment (docprofsky) [reply src/com/teamcautionrobotics/robot2018/Robot.java]: That is not how [`intake.move()`](https://github.com/TeamCautionRobotics/2018-FRC-robot/blob/456c31387e297dd80e6aa24ff4c6b0b2adb8138d/src/com/teamcautionrobotics/robot2018/Intake.java#L39-L47) works.

Comment (docprofsky) [reply src/com/teamcautionrobotics/robot2018/Robot.java]: Though, the bulldoze part might be a problem.

**Review by TaeSeanDo — APPROVED** [review]

**Pull request merged**